drivers/usb/udc/udc_renesas_ra.c

```diff
@@ -0,0 +1,754 @@
+	}
+
+	if (udc_ep_disable_internal(dev, USB_CONTROL_EP_IN)) {
+		LOG_ERR("Failed to disable control endpoint");
+		return -EIO;
+	}
+
+	if (FSP_SUCCESS != R_USBD_Close(&data->udc)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int udc_renesas_ra_driver_preinit(const struct device *dev)
+{
+	const struct udc_renesas_ra_config *config = dev->config;
+	struct udc_data *data = dev->data;
+	uint16_t mps = 1023;
+	int err;
+
+	err = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		return err;
+	}
+
+	k_mutex_init(&data->mutex);
+
+	data->caps.rwup = true;
+	data->caps.mps0 = UDC_MPS0_64;
+	if (config->speed_idx == UDC_BUS_SPEED_HS) {
+		data->caps.hs = true;
+		mps = 1024;
+	}
+
+	for (int i = 0; i < config->num_of_eps; i++) {
+		config->ep_cfg_out[i].caps.out = 1;
+		if (i == 0) {
+			config->ep_cfg_out[i].caps.control = 1;
+			config->ep_cfg_out[i].caps.mps = 64;
+		} else {
+			config->ep_cfg_out[i].caps.bulk = 1;
+			config->ep_cfg_out[i].caps.interrupt = 1;
+			config->ep_cfg_out[i].caps.iso = 1;
+			config->ep_cfg_out[i].caps.mps = mps;
+		}
+
+		config->ep_cfg_out[i].addr = USB_EP_DIR_OUT | i;
+		err = udc_register_ep(dev, &config->ep_cfg_out[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	for (int i = 0; i < config->num_of_eps; i++) {
+		config->ep_cfg_in[i].caps.in = 1;
+		if (i == 0) {
+			config->ep_cfg_in[i].caps.control = 1;
+			config->ep_cfg_in[i].caps.mps = 64;
+		} else {
+			config->ep_cfg_in[i].caps.bulk = 1;
+			config->ep_cfg_in[i].caps.interrupt = 1;
+			config->ep_cfg_in[i].caps.iso = 1;
+			config->ep_cfg_in[i].caps.mps = mps;
+		}
+
+		config->ep_cfg_in[i].addr = USB_EP_DIR_IN | i;
+		err = udc_register_ep(dev, &config->ep_cfg_in[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	config->make_thread(dev);
+	LOG_INF("Device %p (max. speed %d)", dev, config->speed_idx);
+
+	return 0;
+}
+
+static int udc_renesas_ra_lock(const struct device *dev)
+{
+	return udc_lock_internal(dev, K_FOREVER);
+}
+
+static int udc_renesas_ra_unlock(const struct device *dev)
+{
+	return udc_unlock_internal(dev);
+}
+
+static const struct udc_api udc_renesas_ra_api = {
+	.lock = udc_renesas_ra_lock,
+	.unlock = udc_renesas_ra_unlock,
+	.device_speed = udc_renesas_ra_device_speed,
+	.init = udc_renesas_ra_init,
+	.enable = udc_renesas_ra_enable,
+	.disable = udc_renesas_ra_disable,
+	.shutdown = udc_renesas_ra_shutdown,
+	.set_address = udc_renesas_ra_set_address,
+	.host_wakeup = udc_renesas_ra_host_wakeup,
+	.ep_enable = udc_renesas_ra_ep_enable,
+	.ep_disable = udc_renesas_ra_ep_disable,
+	.ep_set_halt = udc_renesas_ra_ep_set_halt,
+	.ep_clear_halt = udc_renesas_ra_ep_clear_halt,
+	.ep_enqueue = udc_renesas_ra_ep_enqueue,
+	.ep_dequeue = udc_renesas_ra_ep_dequeue,
+};
+
+#define DT_DRV_COMPAT renesas_ra_udc
+
+#define USBHS_BASE_ADDR         (0x40252000)
+#define USB_MODULE_NUMBER(addr) ((addr == USBHS_BASE_ADDR) ? 1 : 0)
```

Comment:
I did not understand the reason for this construction.

Reply:
This code is to determine the module number of USB before input to HAL layer. The Renesas RA device has two USB modules, USBFS and USBHS, with module numbers 0 and 1 respectively.

Reply:
If there are different controllers, should not there be two different drivers?

Reply:
There is not much hardware difference between USBHS and USBFS, so they can share the same driver source. Actually, on other platforms (like Renesas RA FSP or TinyUSB library), only 1 driver source is needed for 2 modules.

Reply:
I found out how module_number is used in your HAL.